tycho-ethereum/src/entrypoint_tracer/allowance_slot_detector.rs

```diff
@@ -0,0 +1,904 @@
+                        "Found storage slot for token"
+                    );
+                    token_slots.insert(token.clone(), Ok(slot));
+                }
+                Err(e) => {
+                    error!(token = %token, error = %e, "Failed to extract slot for token");
+                    token_slots.insert(token.clone(), Err(e));
+                }
+            }
+        }
+
+        token_slots
+    }
+
+    /// Extract storage slot from paired debug_traceCall and eth_call responses
+    fn extract_slot_from_paired_responses(
+        &self,
+        token: &Address,
+        debug_response: Option<&Value>,
+        eth_call_response: Option<&Value>,
+    ) -> Result<SlotDetectionResult, AllowanceSlotError> {
+        let debug_resp = debug_response.ok_or_else(|| {
+            AllowanceSlotError::InvalidResponse("Missing debug_traceCall response".into())
+        })?;
+
+        let eth_call_resp = eth_call_response.ok_or_else(|| {
+            AllowanceSlotError::InvalidResponse("Missing eth_call response".into())
+        })?;
+
+        if let Some(error) = debug_resp.get("error") {
+            warn!("Debug trace failed for token {}: {}", token, error);
+            return Err(AllowanceSlotError::RequestError(error.to_string()));
+        }
+
+        if let Some(error) = eth_call_resp.get("error") {
+            warn!("Eth call failed for token {}: {}", token, error);
+            return Err(AllowanceSlotError::RequestError(error.to_string()));
+        }
+
+        // Extract allowance from eth_call response
+        let allowance = self.extract_allowance_from_call_response(eth_call_resp)?;
+
+        // Extract slot values from debug_traceCall response
+        let slot_values = self.extract_slot_values_from_trace_response(debug_resp)?;
+
+        // Find the best slot by comparing values to the expected allowance
+        self.find_best_slot_by_value_comparison(slot_values, allowance)
+    }
+
+    /// Extract allowance from eth_call response
+    fn extract_allowance_from_call_response(
+        &self,
+        response: &Value,
+    ) -> Result<U256, AllowanceSlotError> {
+        let result = response
+            .get("result")
+            .and_then(|v| v.as_str())
+            .ok_or_else(|| {
+                AllowanceSlotError::InvalidResponse("Missing result in eth_call".into())
+            })?;
+
+        let hex_str = result
+            .strip_prefix("0x")
+            .unwrap_or(result);
+        if hex_str.len() != 64 {
+            return Err(AllowanceSlotError::AllowanceExtractionError(format!(
+                "Invalid result length: {} (expected 64)",
+                hex_str.len()
+            )));
+        }
+
+        U256::from_str_radix(hex_str, 16)
+            .map_err(|e| AllowanceSlotError::AllowanceExtractionError(e.to_string()))
+    }
+
+    /// Extract accessed slots with their values from debug_traceCall response
+    fn extract_slot_values_from_trace_response(
```

Comment:
Oh interesting. So we basically make a dummy request and check which slots are accessed? That's not really brute forcing, that's pretty smart? I guess the brute forcing was something else entirely

Reply:
yeees exactly!